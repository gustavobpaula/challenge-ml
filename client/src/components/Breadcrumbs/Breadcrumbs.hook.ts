import { useStore } from 'hooks'

export const useBreadcrumbs = () => {
  const { breadcrumbs } = useStore(state => state)

  return { breadcrumbs }
}
