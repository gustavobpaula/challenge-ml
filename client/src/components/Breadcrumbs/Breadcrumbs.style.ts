import styled from '@emotion/styled'

export const Content = styled.div`
  align-items: center;
  display: flex;
  justify-content: space-between;
`

export const List = styled.ul`
  display: flex;
  list-style: none;
  margin: ${({ theme }) => `${theme.spacings.rule1} 0`};
  padding: 0%;
`

export const ListItem = styled.li`
  color: ${({ theme }) => theme.colors.gray3};
  font-size: ${({ theme }) => theme.type.sizes.xs};

  &::before {
    content: '>';
    margin: 0 5px;
  }

  &:first-child {
    &::before {
      display: none;
    }
  }
`
