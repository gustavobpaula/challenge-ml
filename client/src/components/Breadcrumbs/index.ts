export { default as breadcrumbs } from './Breadcrumbs'
