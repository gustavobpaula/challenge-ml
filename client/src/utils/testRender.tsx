import React from 'react'
import { render, RenderOptions } from '@testing-library/react'
import { ThemeProvider } from '@emotion/react'

import theme from 'config/theme'

type RenderComponentProps = {
  component: React.ReactNode
  useTheme?: boolean
  options?: RenderOptions
}

type ProviderProps = {
  children: React.ReactNode
}

const Theme = ({ children }: ProviderProps) => (
  <ThemeProvider theme={theme}>{children}</ThemeProvider>
)

export const RenderComponent = ({
  component,
  useTheme = false,
  options = {},
}: RenderComponentProps) => {
  const providers = [useTheme && Theme]

  const componentWithProviders = providers.reduce((accComponent, Provider) => {
    if (!Provider) return accComponent

    return <Provider>{accComponent}</Provider>
  }, component)

  return render(<>{componentWithProviders}</>, options)
}
