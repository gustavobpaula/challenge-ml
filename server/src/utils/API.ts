import fetch from 'node-fetch'
import { URLSearchParams } from 'url'

const ML_API = 'https://api.mercadolibre.com'

const API = {
  /**
   * Return currency object
   */
  getCurrency: async (id: string) => {
    try {
      const response = await fetch(`${ML_API}/currencies/${id}`)
      return await response.json()
    } catch (error) {
      throw new Error(error)
    }
  },
  /**
   * Return items by query
   */
  search: async (query: string) => {
    const params = new URLSearchParams()
    params.append('q', query)

    try {
      const response = await fetch(
        `${ML_API}/sites/MLA/search?${params.toString()}`
      )

      return await response.json()
    } catch (error) {
      throw new Error(error)
    }
  },
}

export default API
