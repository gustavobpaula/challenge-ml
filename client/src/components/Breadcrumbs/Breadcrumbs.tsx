import { Container } from 'components'
import { useBreadcrumbs } from './Breadcrumbs.hook'
import * as S from './Breadcrumbs.style'

function Breadcrumbs() {
  const { breadcrumbs } = useBreadcrumbs()
  return (
    <Container>
      <S.List aria-label="breadcrumbs">
        {breadcrumbs.map(breadcrumb => (
          <S.ListItem key={breadcrumb}>{breadcrumb}</S.ListItem>
        ))}
      </S.List>
    </Container>
  )
}

export default Breadcrumbs
