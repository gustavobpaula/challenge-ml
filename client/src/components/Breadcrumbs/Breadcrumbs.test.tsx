import { screen } from '@testing-library/react'

import { RenderComponent } from 'utils'
import Breadcrumbs from './Breadcrumbs'
import * as hooks from './Breadcrumbs.hook'

const itemList = ['item 1', 'item 2', 'item 3']

const makeSut = () => {
  jest.spyOn(hooks, 'useBreadcrumbs').mockImplementation(() => {
    return {
      breadcrumbs: itemList,
    }
  })

  return RenderComponent({
    component: <Breadcrumbs />,
    useTheme: true,
  })
}

describe('Breadcrumb component', () => {
  it('Should render a list', () => {
    makeSut()

    expect(
      screen.getByRole('list', { name: /breadcrumb/i })
    ).toBeInTheDocument()
  })

  it.each(itemList)('Should render a item list %o', item => {
    makeSut()

    expect(screen.getByText(item)).toBeInTheDocument()
  })
})
