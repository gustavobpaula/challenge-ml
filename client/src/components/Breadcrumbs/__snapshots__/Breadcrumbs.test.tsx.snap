// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Breadcrumb component Should render match snapshot 1`] = `
<DocumentFragment>
  <section
    class="css-6hghsz"
  >
    <ul
      aria-label="breadcrumbs"
      class="css-tw8vgv"
    >
      <li
        class="css-ll50vx"
      >
        item 1
      </li>
      <li
        class="css-ll50vx"
      >
        item 2
      </li>
      <li
        class="css-ll50vx"
      >
        item 3
      </li>
    </ul>
  </section>
</DocumentFragment>
`;
